tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/unsafe/UnsafeFunBuilder.scala

```diff
@@ -85,10 +85,27 @@ trait UnsafeFunBuilder extends ProtoBuilder {
     BuilderUtil.composeAndValidateTypes(TlaFunOper.tuple, sig, args: _*)
   }
 
-  /** [x \in S |-> e] */
-  protected def _funDef(e: TlaEx, x: TlaEx, S: TlaEx): TlaEx = {
-    require(x.isInstanceOf[NameEx])
-    buildBySignatureLookup(TlaFunOper.funDef, e, x, S)
+  /** [x1 \in S1, ..., xn \in Sn |-> e], must have at least 1 var-set pair */
+  protected def _funDef(e: TlaEx, varSetPairs: (TlaEx, TlaEx)*): TlaEx = {
+    // the other _funDef does all the require checks
+    val args = varSetPairs.flatMap { case (k, v) =>
+      Seq(k, v)
+    }
+    _funDefMixed(e, args: _*)
+  }
+
+  /**
+   * Alternate call method, where pairs are passed interleaved
+   *
+   * @see
+   *   _funDef[[_funDef(e: TlaEx, varSetPairs: (NameEx, TlaEx)*)]]
+   */
+  protected def _funDefMixed(e: TlaEx, varSetPairs: TlaEx*): TlaEx = {
+    // Even, nonzero # of args and every other argument is NameEx
+    require(varSetPairs.nonEmpty)
+    require(varSetPairs.size % 2 == 0)
```

Comment:
Aren't these already pairs, i.e., an even number of flattened arguments?

Reply:
Yes if chained by other methods (`_funDef`), not necessarily if invoked directly.